### PR TITLE
Add opentelemetry metrics implementation [PLEN-29]

### PR DESCRIPTION
**Pull request by nicu-da**
changelog_begin
changelog_end

### CODE REVIEW

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpentelemetryFactory.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+case class OpentelemetryMeter(name: String, counter: LongCounter) extends Meter {
+  override def mark(value: Long): Unit = counter.add(value)
+}
+
+case class OpentelemetryCounter(name: String, counter: LongUpDownCounter) extends Counter {
+  override def inc(): Unit = counter.add(1)
+  override def inc(n: Long): Unit = counter.add(n)
+  override def dec(): Unit = counter.add(-1)
+  override def dec(n: Long): Unit = counter.add(-n)
+  override def getCount: Long = 0 // Not supported OpenTelemetry
```
Comment (skisel-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpentelemetryFactory.scala:107]:
```suggestion
  override def getCount: Long = 0 // Not supported by OpenTelemetry
```

**Review by skisel-da — APPROVED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import io.opentelemetry.api.metrics.{
+  LongCounter,
+  LongHistogram,
+  LongUpDownCounter,
+  Meter => OtelMeter,
+}
+
+trait OpenTelemetryFactory extends Factory {
+
+  def otelMeter: OtelMeter
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:22]: Nit: can we add newlines between the non-one-liner method definitions, please? It's hard to read the code without the previous.

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      case longInitial: Long =>
+        val varGauge = new VarGauge[Long](longInitial)
+        otelMeter.gaugeBuilder(name).ofLongs().buildWithCallback(_.record(varGauge.getValue))
+        OpentelemetryGauge(name, varGauge.asInstanceOf[VarGauge[T]])
+      case doubleInitial: Double =>
+        val varGauge = new VarGauge[Double](doubleInitial)
+        otelMeter.gaugeBuilder(name).buildWithCallback(_.record(varGauge.getValue))
+        OpentelemetryGauge(name, varGauge.asInstanceOf[VarGauge[T]])
+      case _ =>
+        // A NoOp guage as opentelemetry only supports longs and doubles
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:38]:
```suggestion
        // A NoOp gauge as OpenTelemetry only supports longs and doubles.
```

**Review by miklos-da — APPROVED** [review]
LGTM, only added code readability related suggestions.

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    name,
+    otelMeter.upDownCounterBuilder(name).build(),
+  )
+  override def histogram(name: MetricName): MetricHandle.Histogram = OpentelemetryHistogram(
+    name,
+    otelMeter.histogramBuilder(name).ofLongs().build(),
+  )
+}
+
+case class OpentelemetryTimer(name: String, histogram: LongHistogram) extends Timer {
```
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:74]: That's also an option, it's also included in the package though, not sure how much repetition is too much repetition

**Comment by nicu-da** [comment]
@miklos-da All the comments are fixed here: https://github.com/digital-asset/daml/pull/15357

**Pull request merged**